### ISSUE TITLE
AssertionError: Cannot extract prediction target since the index of forecast is outside the index of target

### ISSUE THREAD

**Issue by dmitra79** [awslabs/gluon-ts#2043]
## Description
 I get an "AssertionError: Cannot extract prediction target since the index of forecast is outside the index of target" when trying to evaluate the prediction (code snippet below).

I probably don't understand something about GluonTS behavior, but if I am feeding in the data to evaluate, shouldn't the forecast dates match the time series by construction?

## To Reproduce
I can't easily create an example or share my data. The problem is possibly with my understanding of GluonTS behavior.
After training a simple predictor, I try to evaluate the forecasts on the test data but get an error:

```
from gluonts.evaluation import make_evaluation_predictions

forecast_it, ts_it = make_evaluation_predictions(
    dataset=tds.test,  
    predictor=predictor,  
    num_samples=100,  
)

forecasts = list(forecast_it)
tss = list(ts_it)

from gluonts.evaluation import Evaluator
evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9], num_workers=None)
agg_metrics, item_metrics = evaluator(iter(tss), iter(forecasts), num_series=len(tds.test))
print(json.dumps(agg_metrics, indent=4))
```

I have checked that the shapes/length are correct:
- tss entries have length 168,
- forecast entries have samples of length 24 (the prediction length I specified when constructing the dataset)

The problem is that the  forecast start doesn't match the time series.  It seems as though perhaps the order of the 
forecast and/or tss gets messed up? This would be easier to debug, btw, if the error message include item_ids and not just the dates.

## Error message or code output

```
Running evaluation:   0%|          | 0/3710 [00:00<?, ?it/s]/home/datascience/conda/derv0/lib/python3.7/site-packages/gluonts/evaluation/_base.py:306: FutureWarning: Timestamp.freq is deprecated and will be removed in a future version
  date_before_forecast = forecast.index[0] - forecast.index[0].freq
Running evaluation:   2%|▏         | 68/3710 [00:00<00:16, 218.64it/s]

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-28-48ee4a1293b7> in <module>
      1 from gluonts.evaluation import Evaluator
      2 evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9], num_workers=None)
----> 3 agg_metrics, item_metrics = evaluator(iter(tss), iter(forecasts), num_series=len(tds.test))
      4 print(json.dumps(agg_metrics, indent=4))
      5 #item_metrics.head()

~/conda/derv0/lib/python3.7/site-packages/gluonts/evaluation/_base.py in __call__(self, ts_iterator, fcst_iterator, num_series)
    226             else:
    227                 for ts, forecast in it:
--> 228                     rows.append(self.get_metrics_per_ts(ts, forecast))
    229 
    230         assert not any(

~/conda/derv0/lib/python3.7/site-packages/gluonts/evaluation/_base.py in get_metrics_per_ts(self, time_series, forecast)
    312         self, time_series: Union[pd.Series, pd.DataFrame], forecast: Forecast
    313     ) -> Mapping[str, Union[float, str, None, np.ma.core.MaskedConstant]]:
--> 314         pred_target = np.array(self.extract_pred_target(time_series, forecast))
    315         past_data = np.array(self.extract_past_data(time_series, forecast))
    316 

~/conda/derv0/lib/python3.7/site-packages/gluonts/evaluation/_base.py in extract_pred_target(time_series, forecast)
    266             forecast.index
    267         ), (
--> 268             "Cannot extract prediction target since the index of forecast is outside the index of target\n"
    269             f"Index of forecast: {forecast.index}\n Index of target: {time_series.index}"
    270         )

AssertionError: Cannot extract prediction target since the index of forecast is outside the index of target
Index of forecast: DatetimeIndex(['2019-06-07 07:00:00', '2019-06-07 08:00:00',
               '2019-06-07 09:00:00', '2019-06-07 10:00:00',
               '2019-06-07 11:00:00', '2019-06-07 12:00:00',
               '2019-06-07 13:00:00', '2019-06-07 14:00:00',
               '2019-06-07 15:00:00', '2019-06-07 16:00:00',
               '2019-06-07 17:00:00', '2019-06-07 18:00:00',
               '2019-06-07 19:00:00', '2019-06-07 20:00:00',
               '2019-06-07 21:00:00', '2019-06-07 22:00:00',
               '2019-06-07 23:00:00', '2019-06-08 00:00:00',
               '2019-06-08 01:00:00', '2019-06-08 02:00:00',
               '2019-06-08 03:00:00', '2019-06-08 04:00:00',
               '2019-06-08 05:00:00', '2019-06-08 06:00:00'],
              dtype='datetime64[ns]', freq='H')
 Index of target: DatetimeIndex(['2020-09-19 07:00:00', '2020-09-19 08:00:00',
               '2020-09-19 09:00:00', '2020-09-19 10:00:00',
               '2020-09-19 11:00:00', '2020-09-19 12:00:00',
               '2020-09-19 13:00:00', '2020-09-19 14:00:00',
               '2020-09-19 15:00:00', '2020-09-19 16:00:00',
               ...
               '2020-09-25 21:00:00', '2020-09-25 22:00:00',
               '2020-09-25 23:00:00', '2020-09-26 00:00:00',
               '2020-09-26 01:00:00', '2020-09-26 02:00:00',
               '2020-09-26 03:00:00', '2020-09-26 04:00:00',
               '2020-09-26 05:00:00', '2020-09-26 06:00:00'],
              dtype='datetime64[ns]', length=168, freq='H')
```

## Environment
    Operating system: Linux
    Python version: 3.7.10
    GluonTS version: 0.9.4
    MXNet version: 1.7.0.post1

Running in a notebook, in a cloud session, on a CPU machine (no GPU).

**Comment by jaheba** [comment]
> I probably don't understand something about GluonTS behavior, but if I am feeding in the data to evaluate, shouldn't the forecast dates match the time series by construction?

Normally, yes.

What is strange here is that the dates of the forecast are in `2019`, while the input data is from `2020`.

What model are you using?

**Comment by dmitra79** [comment]
I've had this with both SimpleFeedForwardEstimator and DeepAREstimator. But I had this work at some point, did some refactoring and got the error. Can't understand how this can ahppen though.

**Comment by dmitra79** [comment]
When I started, I had a small number of very long time series, and on those the code seems to run fine.  I then split these into smaller segments, and at some point started getting this error. I now thought it may be due to multiple time series having the same item_id after splitting, so I integrated start_time (the date) into item_id, ex:
```
> instances_test=[x for x in iter(tds.test)]
> [(x['item_id'],x['start'],len(x['target'])) for x in instances_test[0:5]]

[('9973--2019-08-03', Timestamp('2019-08-03 04:00:00', freq='H'), 168),
 ('9973--2019-08-10', Timestamp('2019-08-10 04:00:00', freq='H'), 168),
 ('9973--2019-08-17', Timestamp('2019-08-17 04:00:00', freq='H'), 168),
 ('9973--2019-08-24', Timestamp('2019-08-24 04:00:00', freq='H'), 168),
 ('9973--2019-08-31', Timestamp('2019-08-31 04:00:00', freq='H'), 168)]
```
However, the errors still occur...

**Comment by jaheba** [comment]
How are you splitting the data?

Can you provide a reproducible example? I suspect that you might have a typo or got the ordering of arguments wrong somewhere. Our evaluation API is a bit rough at the moment, something we want to improve in the future :)

**Comment by dmitra79** [comment]
I am splitting about two years of data into weekly or monthly chunks before converting to GluonTS format. I'll try have a simple reproducible example next week.

**Comment by yuhaod** [comment]
I had the exactly same error when running examples/evaluate_model.py.

**Comment by jaheba** [comment]
That's weird, for me it runs without errors.

What Pandas version are you using?

**Comment by dmitra79** [comment]
I have pandas 1.3.4.  (Can't run the example in my environment due to not having internet access.)

**Comment by yuhaod** [comment]
> 

I have pandas 1.4.2

**Comment by yuhaod** [comment]
Any updates on this? Thanks.

**Comment by jaheba** [comment]
> Any updates on this? Thanks.

I was not able to reproduce the error -- `examples/evaluate_model.py` runs just fine.

Are you running a modified version?

**Comment by yuhaod** [comment]
After updating GluonTS to the latest version. This issue disappears. Thanks!

**Comment by jaheba** [comment]
> After updating GluonTS to the latest version. This issue disappears. Thanks!

Which version did you update to? `0.9.5` or the pre-release of `0.10`?

**Comment by dmitra79** [comment]
I've got a self-contained example. Looks like this happens when `Estimator` has `freq` and `prediction_length` that don't match the data. But this shows up only during evaluation. 

```py
import pandas as pd
import numpy as np
import json
from gluonts.evaluation import Evaluator
from gluonts.evaluation import backtest
from gluonts.dataset.common import ListDataset, TrainDatasets, MetaData
from gluonts.evaluation import make_evaluation_predictions

# from gluonTS tutorial: https://ts.gluon.ai/tutorials/forecasting/extended_tutorial.html#Training-an-existing-model
# create data
N = 10  # number of time series
T = 24 * 7  # number of timesteps
prediction_length = 24
freq = "1H"
custom_dataset = np.random.normal(size=(N, T))
start = pd.Timestamp(
    "01-01-2019", freq=freq
)  # can be different for each time series

# train dataset: cut the last window of length "prediction_length", add "target" and "start" fields
train_ds = ListDataset(
    [
        {"item_id": str(i), "target": x, "start": start}
        for (i, x) in enumerate(custom_dataset[:, :-prediction_length])
    ],
    freq=freq,
)
# test dataset: use the whole dataset, add "target" and "start" fields
test_ds = ListDataset(
    [
        {"item_id": str(i), "target": x, "start": start}
        for (i, x) in enumerate(custom_dataset[:, :])
    ],
    freq=freq,
)

custom_ds_metadata = MetaData(
    target={"name": "target"},
    time_granularity=freq,
    prediction_length=prediction_length,
)

dataset = TrainDatasets(custom_ds_metadata, train_ds, test_ds)


from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
from gluonts.mx import Trainer

estimator = SimpleFeedForwardEstimator(
    num_hidden_dimensions=[10],
    prediction_length=4,  # dataset.metadata.prediction_length,
    context_length=100,
    freq="W",  # dataset.metadata.freq,
    trainer=Trainer(
        ctx="cpu",
        epochs=2,
        learning_rate=1e-3,
        num_batches_per_epoch=100,
    ),
)
predictor = estimator.train(train_ds, test_ds)

evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9], num_workers=None)
back_agg_metrics, back_item_metrics = backtest.backtest_metrics(
    test_ds, predictor, evaluator
)

```

**Comment by lostella** [comment]
@dmitra79 I can confirm that the issue was solved in the `0.10.x` releases: all information about frequency comes now directly from the data (see #1997) and any mis-configuration of the model will result in the model underperforming (if it's configured with the wrong `freq`), or in wrong experimental results (in case the `prediction_length` is set differently from what the experimental data expects).

Also note that in recent releases `SimpleFeedForwardEstimator` does not have a `freq` argument, and that all timestamp in the data are supposed to be `pd.Period` instead of `pd.Timestamp`.
